### ISSUE TITLE
Travis and buildbot integration

### ISSUE THREAD

**Issue by mcallegari** [mcallegari/qlcplus#463]
Discussions started in #459 and #456 to be continued here

**Comment by nomis52** [comment]
Continuing discussion:

>What's the deal with buildbot ? Once configured it starts automatically when the host is online or do I >have to start it manually ?

Usually it's configured to start on boot.

>If so, within the embedded devices that I've got, the Rpi 2 is the one I turn on most frequently.
>As for PCs, Mint is almost always online. The others need to be started "on demand" (even though I find >myself playing pretty frequently on Windows lately... :confounded: )

It's best to only add hosts that are on 24x7, otherwise they have a tendency to get behind on builds, and then you have to remember to go flush the queue.

**Comment by mcallegari** [comment]
@nomis52 then unfortunately I don't have 24/7 devices online. If you want though I can offer my fun park machines ( :smile: ) as "medium rate" contributing slaves.
I've seen in your waterfall that there are already some not-always-online slaves. Am I correct ?

@peternewman I think I managed to silence the false-positive results of Travis with this 410c3c26800c3f69da430ab6e057b48723dcbe58 and this 3b2ed24d6e232dc0ea0b7e6bc05e193a87857313
Basically the timer test is not performed when building with Travis

I've seen that the qlcplus buildbot has been setup ! Many thanks for that !
Where can I monitor the builds it produces ?

**Comment by nomis52** [comment]
Alright: https://qlc-buildbot.openlighting.org/ is up and running. We'll need to setup the slaves now.

What do you mean by "medium rate" machines?

**Comment by mcallegari** [comment]
> What do you mean by "medium rate" machines?

I mean machines not 24/7 online but upon request, maybe 6-7 hours a week ?
If you let me know on which of the devices I mentioned you are interested in, I'll make sure to setup buildbot on them and get them online a decent number of times per week

**Comment by peternewman** [comment]
@nomis52 can we move it to https://qlcplus-buildbot.openlighting.org/ just in case... and to match the config/repo.

**Comment by nomis52** [comment]
Done.

@mcallegari it really depends on:
i) how many changes you have
ii) the time it takes for each build

Lets add some of your slaves and see how they go. Choose the two that would be the most useful and follow the instructions on https://wiki.openlighting.org/index.php/OLA_Buildbot
For QLC+ use buildmaster.openlighting.org:9990

**Comment by peternewman** [comment]
@nomis52 if we add the is_slow option to @mcallegari 's bots, they'll have fewer builds to do and may manage to keep up hopefully.

@mcallegari I think I have a slightly nicer workaround for Travis; expect a PR shortly.

@mcallegari I've now got our Ubuntu slave working fine, and the Mac one is almost there, apart from it's failing one test:QWARN  : InputOutputMap_Test::profileDirectories() QCoreApplication::applicationDirPath: Please instantiate the QApplication object first
FAIL!  : InputOutputMap_Test::profileDirectories() 'dir.exists() == true' returned FALSE. ()
   Loc: [inputoutputmap_test.cpp(491)]

**Comment by peternewman** [comment]
Digging further @mcallegari that test seems to be failing because Library/Application Support/QLC+ doesn't exist, but it's seemingly not trying to create it, which seems to me to be a slightly flawed test. Although I appreciate running the test without running the app is a bit odd, I could see people doing make && make check && make install before running it, but at that point the tests will fail because they've not run it.

**Comment by mcallegari** [comment]
@nomis52 Ok, I'll set up my raspberry 2 and my Mint 17 as soon as I have some time

@peternewman Thanks for spotting the test. I think it's simply because when you develop QLC+ you surely did make install and ran the app before running the test units :smile: 
On OSX the create-dmg script doesn't actually run the test units. They are checked only during the Linux build, which is normally the first that I pack.
I would say we can add this case to the "IS_TRAVIS" workaround && "are we on OSX" or eventually don't even run the tests on OSX

**Comment by peternewman** [comment]
@mcallegari I'd say the reverse actually, there's no point installing something that's broken. So if I'm going to bother running the tests at all, I'd make check, then make install.

Also regarding your dmg script, aside from/before Buildbot, how did you know there weren't Mac specific bugs/test failures if the unit tests are only run on Linux? Personally I'd say it should be run on all platforms where possible.

Not running the tests on OSX via Buildbot completely misses the point of unit tests and Buildbot. All that means you're doing then with Buildbot is checking there are no syntax/complilation errors against the OSX libraries. I'm guessing I can fix it by just creating that folder or similar, but I'd still be inclined to fix the test somehow (work in a scratch folder, create the real folder or something similar).

**Comment by mcallegari** [comment]
@peternewman you are absolutely right, but please consider that I didn't have a real Mac until the other day.
Before I was using a crappy iAtkos image, without the proper hardware support and many other limitations.
So I didn't even attempt to run test units on such an unreliable system. I've always relied on users feedbacks.
It's a fact though that I've done 28 releases in this way and I can't say OSX users are unhappy :wink: 

Now a few things have changed: I've got a Mac Mini, there are Travis and Buildbot. Please bare with me in these settlement days :smiley: 

I'll sit down on my Mac and review the test units on OSX. I agree that some tests must be refined.

**Comment by mcallegari** [comment]
@peternewman so, I've started my adventure into buildbot slavery :smile: 
I'm a bit confused by test units though:

1- I don't understand why Travis build is OK while this slave fails:
https://qlcplus-buildbot.openlighting.org/builders/buildcheck-qlcplus-ubuntu-tahr-i686-noopenslp/builds/33/steps/unit%20tests/logs/stdio

2- I've setup a slave myself and it fails at the first test requiring a running X server. Are you aware of any solution/workaround to solve this ?
https://qlcplus-buildbot.openlighting.org/builders/buildcheck-qlcplus-mint-rebecca-x86_64-1/builds/1/steps/unit%20tests/logs/stdio

3- I've had a look at the test failing on OSX. Actually the underneath code is supposed to create the requested directory. See here:
https://github.com/mcallegari/qlcplus/blob/master/engine/src/qlcfile.cpp#L246
I'll try it manually on my Mac to see if something clearly wrong is going on.

**Comment by mcallegari** [comment]
Yeah ! After 5702ac6af7dd006b2b6d8232bedee855ef04fc2b my Linux slave reaches the end successfully.
I am now setting up my RPi 2 slave as well. Will be online soon.

Next step is fixing the unit test(s) failing on OSX.

@nomis52 I noticed that your ubuntu-tahr slave is failing the timer unit test. Are you running it on a virtual machine ? If so, how can we fix that in this case ? Is it even possible to detect that we're running on a VM ?

**Comment by nomis52** [comment]
It's not on a VM but it is a fairly old machine:

Athlon(tm) XP 2200+
MemTotal:  474748 kB
